<h1 style="color:#1565C0;">Week 4 - Day 1</h1>

<h2 style="color:#6A1B9A;">Train, Validation and Test Sets</h2>

<h3 style="color:#00897B;">Breast Cancer Classification</h3>

This notebook continues the classification work from Week 3 using the Breast Cancer Wisconsin dataset.

The goal today is not to build a new classification system, but to improve the way I evaluate and select models.

In [22]:
import pandas as pd

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

In [23]:
data = load_breast_cancer()

X = pd.DataFrame(
    data.data,
    columns=data.feature_names
)

y = pd.Series(
    (data.target == 0).astype(int),
    name="Malignant"
)

X.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


<h2 style="color:#6A1B9A;">Dataset Reminder</h2>

This is the same dataset used during Week 3 for supervised classification.

The target is defined as:

- <span style="color:#C62828;"><b>1 → Malignant</b></span>
- <span style="color:#2E7D32;"><b>0 → Benign</b></span>

The dataset contains numerical measurements extracted from breast mass images.

Before splitting the data, I will check its size and class distribution.

In [24]:
print("X shape:", X.shape)
print("y shape:", y.shape)

print("\nTarget distribution:")
print(y.value_counts())

print("\nTarget percentage:")
print(y.value_counts(normalize=True) * 100)

X shape: (569, 30)
y shape: (569,)

Target distribution:
Malignant
0    357
1    212
Name: count, dtype: int64

Target percentage:
Malignant
0    62.741652
1    37.258348
Name: proportion, dtype: float64


<h2 style="color:#6A1B9A;">Why Use Train, Validation and Test Sets?</h2>

In Week 3, I mainly used training and test sets.

At first, it may seem unnecessary to remove part of the training data and use it as a validation set, because comparing models on the test set does not directly pass the test samples into `model.fit()`.

However, repeatedly checking the test score still affects my decisions.

For example, if I compare several models or several hyperparameter values using the test set, I may choose one specifically because it performed well on that particular test sample.

Therefore:

- <span style="color:#1565C0;"><b>Training set</b></span> → trains the model.
- <span style="color:#EF6C00;"><b>Validation set</b></span> → guides my modeling decisions.
- <span style="color:#C62828;"><b>Test set</b></span> → gives the final independent evaluation.

<span style="color:#00897B;"><b>Key idea:</b></span>  
The training set trains the model, while the validation set guides my decisions.

Even if the test set is never used inside `.fit()`, repeatedly using its results to choose models or hyperparameters makes the test set influence the development process.

<h2 style="color:#6A1B9A;">Three-Way Split</h2>

I will divide the dataset into:

- <span style="color:#1565C0;"><b>60% Training</b></span>
- <span style="color:#EF6C00;"><b>20% Validation</b></span>
- <span style="color:#C62828;"><b>20% Test</b></span>

Because this is a classification problem, I will use `stratify` to preserve approximately the same class distribution in each subset.

In [25]:
X_temp, X_test, y_temp, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

In [26]:
X_train, X_val, y_train, y_val = train_test_split(
    X_temp,
    y_temp,
    test_size=0.25,
    random_state=42,
    stratify=y_temp
)

In [27]:
print("Train:", X_train.shape, y_train.shape)
print("Validation:", X_val.shape, y_val.shape)
print("Test:", X_test.shape, y_test.shape)

print("\nTrain distribution:")
print(y_train.value_counts())

print("\nValidation distribution:")
print(y_val.value_counts())

print("\nTest distribution:")
print(y_test.value_counts())

Train: (341, 30) (341,)
Validation: (114, 30) (114,)
Test: (114, 30) (114,)

Train distribution:
Malignant
0    214
1    127
Name: count, dtype: int64

Validation distribution:
Malignant
0    71
1    43
Name: count, dtype: int64

Test distribution:
Malignant
0    72
1    42
Name: count, dtype: int64


<h3 style="color:#00897B;">Split Results</h3>

The dataset was successfully divided into approximately:

- <span style="color:#1565C0;"><b>Training:</b></span> 341 samples
- <span style="color:#EF6C00;"><b>Validation:</b></span> 114 samples
- <span style="color:#C62828;"><b>Test:</b></span> 114 samples

The original dataset contains approximately **37.3% malignant cases**.

After using `stratify`, the malignant class remained close to this proportion in all three subsets:

- Training: 127 malignant cases
- Validation: 43 malignant cases
- Test: 42 malignant cases

<span style="color:#00897B;"><b>Observation:</b></span>  
`stratify` preserved the class distribution while the data was divided into separate training, validation, and test roles.

<span style="color:#C62828;"><b>From this point, the test set will remain unused until the final evaluation.</b></span>

<h2 style="color:#6A1B9A;">KNN Validation Experiment</h2>

In Week 3, I experimented with the K-Nearest Neighbors classifier and compared different values of `k`.

I will repeat this experiment using the new three-way split.

The purpose is different now:

<span style="color:#EF6C00;"><b>I will select `k` using the validation set instead of the test set.</b></span>

Because KNN is distance-based, the numerical features must also be scaled before training.

In [28]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)

<h3 style="color:#00897B;">Validation Strategy</h3>

For every value of `k`:

1. The KNN model will be trained using only the training set.
2. Its F1-score will be measured on the validation set.
3. The value of `k` with the best validation F1-score will be selected.

<span style="color:#C62828;"><b>The test set will not be checked while selecting `k`.</b></span>

In [29]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score

k_values = [1, 3, 5, 7, 11, 21]

results = []

for k in k_values:
    model = KNeighborsClassifier(n_neighbors=k)

    model.fit(X_train_scaled, y_train)

    train_pred = model.predict(X_train_scaled)
    val_pred = model.predict(X_val_scaled)

    train_f1 = f1_score(y_train, train_pred)
    val_f1 = f1_score(y_val, val_pred)

    results.append([k, train_f1, val_f1])

results_df = pd.DataFrame(
    results,
    columns=["K", "Train F1", "Validation F1"]
)

results_df

,K,Train F1,Validation F1
0,1,1.000000,0.933333
1,3,0.971888,0.952381
2,5,0.963855,0.963855
3,7,0.963855,0.963855
4,11,0.947368,0.963855
5,21,0.925000,0.951220


<h3 style="color:#00897B;">KNN Validation Results</h3>

The highest validation F1-score was:

<span style="color:#EF6C00;"><b>Validation F1 = 0.964</b></span>

Three values achieved the same best validation score:

- `k = 5`
- `k = 7`
- `k = 11`

The results also show that `k = 1` achieved a perfect training F1-score but a lower validation score. This means that excellent training performance does not necessarily mean better performance on unseen data.

<span style="color:#C62828;"><b>Important:</b></span>  
I should not test `k = 5`, `k = 7`, and `k = 11` on the test set and then choose the best one, because that would make the test set part of the tuning process.

<h3 style="color:#00897B;">Final Model Selection</h3>

The best validation F1-score was approximately **0.964**, achieved by:

- `k = 5`
- `k = 7`
- `k = 11`

Since these values are tied on the validation set, I will use:

<span style="color:#1565C0;"><b>k = 7</b></span>

as a middle value among the tied candidates.

<span style="color:#EF6C00;"><b>Important:</b></span>  
The validation results do not prove that `k = 7` is better than `k = 5` or `k = 11`. It is simply a tie-breaking choice made before looking at the test set.

I will not use the test set to break this tie, because doing so would allow the test results to influence model selection and would reduce the reliability of the final evaluation.

In [30]:
X_test_scaled = scaler.transform(X_test)

In [31]:
final_model = KNeighborsClassifier(n_neighbors=7)

final_model.fit(X_train_scaled, y_train)

test_pred = final_model.predict(X_test_scaled)

test_f1 = f1_score(y_test, test_pred)

print("Selected k:", 7)
print("Validation F1:", results_df["Validation F1"].max())
print("Final Test F1:", test_f1)

Selected k: 7
Validation F1: 0.963855421686747
Final Test F1: 0.925


<h2 style="color:#6A1B9A;">Final Test Evaluation</h2>

After selecting `k = 7` using the validation set, I evaluated the final KNN model on the test set for the first and final time.

The results were:

- <span style="color:#EF6C00;"><b>Validation F1:</b></span> 0.964
- <span style="color:#C62828;"><b>Final Test F1:</b></span> 0.925

The test score is lower than the validation score. This is expected because the validation and test sets contain different unseen samples.

<span style="color:#00897B;"><b>Key point:</b></span>  
The test score is used to report the final performance, not to make another modeling decision.

Therefore, I will not change `k` based on the test result.

<h2 style="color:#6A1B9A;">Why Should I Not Tune Using the Test Set?</h2>

Using the test set during model selection would not necessarily mean that the test samples were directly used inside `model.fit()`.

However, the test results would still influence my decisions.

For example, if I tested several values of `k` on the test set and selected the one with the highest test F1-score, my final choice would become biased toward that specific test sample.

This would reduce the reliability of the test score as an estimate of performance on truly unseen data.

<span style="color:#00897B;"><b>My main observation:</b></span>  
The problem is not only direct data leakage. The problem is also allowing test results to influence model-selection decisions.

Therefore:

- <span style="color:#1565C0;"><b>Training data</b></span> trains the model.
- <span style="color:#EF6C00;"><b>Validation data</b></span> is allowed to influence model selection and tuning.
- <span style="color:#C62828;"><b>Test data</b></span> should remain independent until the final evaluation.

Once the test set is evaluated, I should report the result rather than continue tuning the model based on it.

<h2 style="color:#1565C0;">Day 1 Conclusion</h2>

In this notebook, I reused the Breast Cancer classification dataset from Week 3 and changed the evaluation workflow from a simple train/test split to a three-way split.

### Workflow

**Dataset → Train / Validation / Test → Tune on Validation → Final Test**

I used:

- 60% of the data for training.
- 20% for validation.
- 20% for final testing.
- `stratify` to preserve the class distribution.
- `StandardScaler` fitted only on the training data.
- KNN as the Week 3 model.
- Validation F1-score to tune `n_neighbors`.

The best validation F1-score was approximately **0.964**, with `k = 5`, `k = 7`, and `k = 11`.

I selected `k = 7` as a middle value among the tied candidates without using the test set to break the tie.

The final model achieved:

<span style="color:#C62828;"><b>Final Test F1 = 0.925</b></span>

<span style="color:#00897B;"><b>Main lesson:</b></span>  
A trustworthy test score requires keeping the test set independent not only from model training, but also from the decisions made during model development.

<span style="color:#EF6C00;"><b>Remaining limitation:</b></span>  
The model selection still depends on one validation split. A different validation sample could produce different results. This motivates cross-validation in the next step.